File: rank-fixes.mjs
#!/usr/bin/env node
// Build fix-plan.json from already-produced audit artifacts.
//
// JS owns artifact loading/writing and package.json public deep-import risk
// discovery. lumin-audit-core owns finding flattening, tier classification,
// support evidence projection, sorting, summary, and safeFixGroups.

import { writeFileSync } from 'node:fs';
import path from 'node:path';
import { parseCliArgs } from './_lib/cli.mjs';
import { loadIfExists as loadArtifact } from './_lib/artifacts.mjs';
import { runAuditCoreJsonResultFile } from './_lib/audit-core.mjs';
import {
  findNearestPackageInfo,
  getPublicDeepImportRisk,
} from './_lib/package-exports.mjs';

const RANK_FIXES_REQUEST_SCHEMA_VERSION = 'lumin-rank-fixes-producer-request.v1';

const { root, output } = parseCliArgs();
const ROOT = path.resolve(root);
const OUT = path.resolve(output);

const loadIfExists = (name) => loadArtifact(OUT, name, { tag: 'rank-fixes' });

const deadClassify = loadIfExists('dead-classify.json');
if (!deadClassify) {
  console.error('[rank-fixes] dead-classify.json is required. Run classify-dead-exports.mjs first.');
  process.exit(1);
}

const runtimeEvidence = loadIfExists('runtime-evidence.json');
const staleness = loadIfExists('staleness.json');
const symbols = loadIfExists('symbols.json');
const exportActionSafety = loadIfExists('export-action-safety.json');
const callGraph = loadIfExists('call-graph.json');
const entrySurface = loadIfExists('entry-surface.json');
const moduleReachability = loadIfExists('module-reachability.json');

function normalizeRel(file) {
  return String(file ?? '').replace(/\\/g, '/').replace(/^\.\//, '');
}

function addFindingFiles(files, list) {
  for (const item of list ?? []) {
    if (typeof item?.file === 'string' && item.file.length > 0) {
      files.add(normalizeRel(item.file));
    }
  }
}

function collectDeadClassifyFiles(artifact) {
  const files = new Set();
  addFindingFiles(files, artifact?.proposal_C_remove_symbol);
  addFindingFiles(files, artifact?.proposal_A_demote_to_internal);
  addFindingFiles(files, artifact?.proposal_B_review);
  addFindingFiles(files, artifact?.proposal_remove_export_specifier);
  addFindingFiles(files, artifact?.proposal_DEGRADED_unprocessed);
  addFindingFiles(files, artifact?.excludedCandidates);
  return [...files].sort();
}

function publicDeepImportRiskForFile(file) {
  const packageInfo = findNearestPackageInfo(ROOT, file);
  if (!packageInfo?.packageJson) {
    return { risk: false, reason: 'package-json-absent', relFileFromPkgRoot: file };
  }
  return getPublicDeepImportRisk(packageInfo.packageJson, packageInfo.relFileFromPkgRoot);
}

function buildPublicDeepImportRiskByFile(deadClassifyArtifact) {
  const entries = {};
  for (const file of collectDeadClassifyFiles(deadClassifyArtifact)) {
    entries[file] = publicDeepImportRiskForFile(file);
  }
  if (Object.keys(entries).length === 0) {
    entries.__lumin_empty_fix_plan__ = { risk: false, reason: 'empty-fix-plan' };
  }
  return entries;
}

const inputs = {
  'dead-classify.json': true,
  'runtime-evidence.json': !!runtimeEvidence,
  'staleness.json': !!staleness,
  'symbols.json': !!symbols,
  'export-action-safety.json': !!exportActionSafety,
  'call-graph.json': !!callGraph,
  'entry-surface.json': !!entrySurface,
  'module-reachability.json': !!moduleReachability,
};

const artifact = runAuditCoreJsonResultFile(
  ['rank-fixes-artifact', '--input', '-'],
  'rank-fixes-artifact',
  {
    input: JSON.stringify({
      schemaVersion: RANK_FIXES_REQUEST_SCHEMA_VERSION,
      root: ROOT,
      generated: new Date().toISOString(),
      artifacts: {
        deadClassify,
        runtimeEvidence,
        staleness,
        symbols,
        exportActionSafety,
        callGraph,
        entrySurface,
        moduleReachability,
      },
      publicDeepImportRiskByFile: buildPublicDeepImportRiskByFile(deadClassify),
    }),
  }
);

const outPath = path.join(OUT, 'fix-plan.json');
writeFileSync(outPath, JSON.stringify(artifact, null, 2));

console.log('\n══════ fix-plan ranking ══════');
console.log(`  SAFE_FIX    : ${artifact.summary?.SAFE_FIX ?? 0}  (auto-fix candidates)`);
console.log(`  REVIEW_FIX  : ${artifact.summary?.REVIEW_FIX ?? 0}  (human review recommended)`);
console.log(`  DEGRADED    : ${artifact.summary?.DEGRADED ?? 0}  (evidence insufficient — not a hard warning)`);
console.log(`  MUTED       : ${artifact.summary?.MUTED ?? 0}  (policy-excluded — not a finding)`);
console.log(`  total       : ${artifact.summary?.total ?? 0}`);

const resolver = artifact.meta?.resolverBlindness;
const degradedByUnresolvedSpec = (artifact.degraded ?? []).filter((entry) =>
  typeof entry?.reason === 'string' && entry.reason.startsWith('unresolved-spec-could-match')
).length;
if (resolver && resolver.gate === 'tripped') {
  console.log(`\n  ⚠ resolver unresolvedRatio = ${(resolver.ratio * 100).toFixed(1)}%`);
  if (degradedByUnresolvedSpec > 0) {
    console.log(`    ${degradedByUnresolvedSpec} finding(s) DEGRADED by per-finding spec match — add a tsconfig path or alias to reduce.`);
  } else {
    console.log('    No finding matched an unresolved specifier locally; global ratio is informational only.');
  }
}

if ((artifact.summary?.SAFE_FIX ?? 0) > 0) {
  console.log('\n── SAFE_FIX top entries ──');
  for (const entry of (artifact.safeFixes ?? []).slice(0, 5)) {
    console.log(`  ${entry.finding.file}:${entry.finding.line}  ${entry.finding.symbol}  (${entry.reason})`);
  }
  if ((artifact.safeFixGroups ?? []).length > 0) {
    console.log('\n── SAFE_FIX grouped patterns ──');
    for (const group of artifact.safeFixGroups.slice(0, 5)) {
      const sample = group.symbols.slice(0, 4).join(', ');
      const suffix = group.symbols.length > 4 ? ', ...' : '';
      console.log(`  ${group.count}×  ${group.actionKind}  ${group.file}  (${sample}${suffix})`);
    }
  }
}

console.log(`\n[rank-fixes] saved → ${outPath}`);

File: skills/lumin-repo-lens-lab/_engine/rust/rust-main/lumin-audit-core/src/rank_fixes.rs
use anyhow::{bail, Result};
use serde_json::{json, Value};

mod evidence;
mod findings;
mod policy;
mod projection;
mod protocol;
mod support;

use evidence::{
    evidence_for_finding, resolver_blindness_meta, resolver_summary, runtime_by_key,
    staleness_by_key,
};
use findings::{action_by_id, excluded_findings, merge_action_evidence, ordinary_findings};
use policy::{tier_for_finding, Tier};
use projection::{build_safe_fix_groups, by_tier, scored_entry, sort_tier_entries, summary};
pub use protocol::{
    PublicDeepImportRisk, RankFixesArtifact, RankFixesArtifacts, RankFixesRequest,
    RANK_FIXES_REQUEST_SCHEMA_VERSION,
};
use support::with_evidence_support;

const TOOL_NAME: &str = "rank-fixes.mjs";

pub fn build_rank_fixes_artifact(request: RankFixesRequest) -> Result<RankFixesArtifact> {
    if request.schema_version != RANK_FIXES_REQUEST_SCHEMA_VERSION {
        bail!(
            "rank-fixes-artifact: unsupported schemaVersion '{}'",
            request.schema_version
        );
    }
    if request.public_deep_import_risk_by_file.is_empty() {
        bail!("rank-fixes-artifact: missing publicDeepImportRiskByFile");
    }

    let runtime_by = runtime_by_key(request.artifacts.runtime_evidence.as_ref());
    let staleness_by = staleness_by_key(request.artifacts.staleness.as_ref());
    let action_by_id = action_by_id(request.artifacts.export_action_safety.as_ref());
    let resolver = resolver_summary(request.artifacts.symbols.as_ref());

    let mut findings = ordinary_findings(&request.artifacts.dead_classify);
    for finding in &mut findings {
        merge_action_evidence(finding, &action_by_id);
    }
    let muted_findings = excluded_findings(&request.artifacts.dead_classify);

    let mut scored = Vec::new();
    for finding in findings {
        let ranked_finding = with_evidence_support(finding.value.clone(), &request);
        let runtime = runtime_by.get(&finding.key);
        let staleness = staleness_by.get(&finding.key);
        let evidence =
            evidence_for_finding(&finding, runtime, staleness, resolver.as_ref(), &request);
        let result = tier_for_finding(&ranked_finding, &evidence);
        scored.push(scored_entry(ranked_finding, evidence, result));
    }

    for finding in muted_findings {
        let evidence = json!({
            "policy": {
                "excluded": true,
                "reason": finding.excluded_reason.as_deref().unwrap_or("unknown")
            }
        });
        let result = tier_for_finding(&finding.value, &evidence);
        scored.push(scored_entry(finding.value, evidence, result));
    }

    let mut by_tier = by_tier(scored);
    sort_tier_entries(&mut by_tier);
    let safe_fix_groups = build_safe_fix_groups(by_tier.get(&Tier::SafeFix));
    let summary = summary(&by_tier, safe_fix_groups.len());
    let inputs = input_summary(&request);

    Ok(RankFixesArtifact {
        meta: json!({
            "generated": request.generated,
            "root": request.root,
            "tool": TOOL_NAME,
            "executionOwner": "lumin-audit-core",
            "inputs": inputs,
            "resolverBlindness": resolver_blindness_meta(resolver.as_ref()),
            "topUnresolvedSpecifiers": request.artifacts.symbols
                .as_ref()
                .and_then(|symbols| symbols.get("topUnresolvedSpecifiers"))
                .cloned()
                .unwrap_or_else(|| json!([])),
        }),
        summary,
        safe_fixes: by_tier.remove(&Tier::SafeFix).unwrap_or_default(),
        safe_fix_groups,
        review_fixes: by_tier.remove(&Tier::ReviewFix).unwrap_or_default(),
        degraded: by_tier.remove(&Tier::Degraded).unwrap_or_default(),
        muted: by_tier.remove(&Tier::Muted).unwrap_or_default(),
    })
}

fn input_summary(request: &RankFixesRequest) -> Value {
    json!({
        "dead-classify.json": true,
        "runtime-evidence.json": request.artifacts.runtime_evidence.is_some(),
        "staleness.json": request.artifacts.staleness.is_some(),
        "symbols.json": request.artifacts.symbols.is_some(),
        "export-action-safety.json": request.artifacts.export_action_safety.is_some(),
        "call-graph.json": request.artifacts.call_graph.is_some(),
        "entry-surface.json": request.artifacts.entry_surface.is_some(),
        "module-reachability.json": request.artifacts.module_reachability.is_some()
    })
}
